sdk/clientcore/core/src/main/java/io/clientcore/core/implementation/instrumentation/fallback/FallbackAttributes.java

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package io.clientcore.core.implementation.instrumentation.fallback;
+
+import io.clientcore.core.instrumentation.InstrumentationAttributes;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+class FallbackAttributes implements InstrumentationAttributes {
+    private final Map<String, Object> attributes;
+
+    FallbackAttributes(Map<String, Object> attributes) {
+        if (attributes == null) {
+            this.attributes = Collections.emptyMap();
+            return;
+        }
+
+        for (Map.Entry<String, Object> entry : attributes.entrySet()) {
+            Objects.requireNonNull(entry.getKey(), "attribute key cannot be null.");
+            Objects.requireNonNull(entry.getValue(), "attribute value cannot be null.");
+        }
+        this.attributes = Collections.unmodifiableMap(attributes);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public InstrumentationAttributes put(String key, Object value) {
+        Objects.requireNonNull(key, "'key' cannot be null.");
+        Objects.requireNonNull(value, "'value' cannot be null.");
+
+        Map<String, Object> newAttributes = new HashMap<>(attributes.size() + 1);
+        newAttributes.putAll(attributes);
+        newAttributes.put(key, value);
```

Comment:
nit: Based on the default capacity before regrowth the HashMap we're creating it going to regrow right away, before the call to put. Default behavior is at 75% capacity to regrow, we'll want to use some better math here to prevent that.

`new HashMap<>((attributes.size() + 1) * 1.5)`